project3/coordinator/message_log.cpp

```diff
@@ -0,0 +1,37 @@
+/**
+ * @file Implementation of MessageLog.
+ */
+#include "message_log.h"
+namespace coordinator::message_log {
+
+bool MessageLog::Message::operator==(const Message &message) const {
+  return timestamp == message.timestamp && msg == message.msg &&
+         participant_id == message.participant_id;
+}
+
+size_t MessageLog::Hash::operator()(const Message &message) const {
+  // ensure uniqueness from this messages length and time stamp
+  return std::hash<int>()(message.msg.length() + message.timestamp);
```

Comment:
Do we want to include the participant ID in this hash as well? In general, it is good practice to make the hashing and equality operations correspond exactly to each-other.